[FLINK-16437][runtime] Make SlotManager allocate resource from ResourceManager at the worker granularity.

xintongsong opened the pull request:
## What is the purpose of the change

This is the first step of FLINK-14106, including all the major changes inside SlotManager and changes to the RM/SM interfaces, except changes for metrics and status.

At the end of this step, SlotManager should allocate resource from ResourceManager with a WorkerResourceSpec, instead of slot ResourceProfile. At this step, the WorkerResourceSpec will not be used, and the active RMs will always use `ActiveResourceManager#taskExecutorProcessSpec` for requesting TMs. We will change that in subsequent steps.

## Brief change log

- 5ee2a8ece3eefb4726618ed31a05a2e11737b57a..7d6bc425928a3b4f814ebd9a1a7dd19adc5a497c: Minor code clean-ups.
- 8f43c28607d29a23e86873ee7744cb54d9c356e5: Introduce `WorkerResourceSpec`.
- be5aa79fe00d32c91a953cfa6bc7e93b99807599: Create `SlotManagerImpl` with default `WorkerResourceSpec` in active resource manager setups.
- 0f77b1e538f096e75ddc9cea253ae77abd8e08d3: Create SlotManagerImpl with default numSlotsPerWorker.
- a88d719456bd225ffe7b5f95acb951fdfffe74dc: Compute pending slot profiles inside SlotManager when allocating resource.
  - This means also check unfulfillable slot profiles inside SlotManager.
- a4d73c5fcb893ed671f1f058b55f1d1a2cc46583: ResourceManager retrieve a collection of pending workers from SlotManager, instead of number of pending slots.
- 168639b5d6f1035dea8a8a7330de4416c089dbb5: Remove numSlotsPerTaskManager from ActiveResourceManager and ContaineredTaskManagerParameters.
  - Since now we compute pending slot profiles inside SlotManager, ResourceManagers no longer need to be aware of number of slots per worker.
- df11b25dc9094509f24cec5484a6c5b01622d3f8: SlotManager allocate resource from ResourceManager with WorkerRequest instead of ResourceProfile.
  - This marks ResourceManager is no longer aware of slot ResourceProfiles.

## Verifying this change

This is a refactoring work. Most of the behaviors are already covered by exist test cases. Only added a few new test cases for the added data structures and logics.
- Add `WorkerResourceSpecTest`.
- Add test case in `TaskExecutorProcessUtilsTest`.
- Add test case in `SlotManagerImplTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnResourceManagerFactory.java:
```diff
@@ -81,4 +91,27 @@ public static YarnResourceManagerFactory getInstance() {
 	}
+
+	@Override
+	protected CPUResource getDefaultCpus(final Configuration configuration) {
```
Yes, I'm aware of this.
It's just we are currently having both default `TaskExecutorProcessSpec` and default `WorkerResourceSpec`, and will migrate from the former to the latter for the active RMs one by one. The `ActiveResourceManager#getCpuCores` could be removed soon after all active RMs are migrated. That's why I didn't bother to de-duplicate them.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesResourceManagerFactory.java:
```diff
@@ -80,4 +84,10 @@ public static KubernetesResourceManagerFactory getInstance() {
 	}
+
+	@Override
+	protected CPUResource getDefaultCpus(Configuration configuration) {
```
Same here.

xintongsong commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerFactory.java:
```diff
@@ -98,4 +101,10 @@ public MesosResourceManagerFactory(@Nonnull MesosServices mesosServices, @Nonnul
 	}
+
+	@Override
+	protected CPUResource getDefaultCpus(Configuration configuration) {
```
Same here.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerFactory.java:
```diff
@@ -81,4 +83,14 @@ private static Configuration createActiveResourceManagerConfiguration(Configurat
 		@Nullable String webInterfaceUrl,
 		ResourceManagerMetricGroup resourceManagerMetricGroup) throws Exception;
+
+	protected WorkerResourceSpec createDefaultWorkerResourceSpec(Configuration configuration) {
```
I tend to disagree, for the following reasons.
- The default `WorkerResourceSpec` is needed for both static and dynamic slot allocation. For dynamic allocation, we need it for scenarios where some jobs do not provide fine grained resource requirement.
- We need the default `WorkerResourceSpec` only for active RMs, but not for standalone RM. SM does not know which RM it runs on.
- We need the default CPU cores for creating the default `WorkerResourceSpec`, which is depending on different configuration options per active RM. SM does not have that information.
- According to the dependency injection principle, we should try to avoid creating dependencies inside the component if possible.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerConfiguration.java:
```diff
@@ -43,19 +46,23 @@
 			boolean waitResultConsumedBeforeRelease,
-			boolean evenlySpreadOutSlots) {
+			boolean evenlySpreadOutSlots,
+			@Nullable WorkerResourceSpec defaultWorkerResourceSpec) {
```
The default WorkerResourceSpec is needed for both static and dynamic slot allocation. For dynamic allocation, we need it for scenarios where some jobs do not provide fine grained resource requirement.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -125,20 +126,25 @@
 		this.slotRequestTimeout = Preconditions.checkNotNull(slotRequestTimeout);
 		this.taskManagerTimeout = Preconditions.checkNotNull(taskManagerTimeout);
 		this.waitResultConsumedBeforeRelease = waitResultConsumedBeforeRelease;
+		this.workerResourceSpec = workerResourceSpec;
```
See my reply above.

xintongsong commented:
Thanks for the review, @KarmaGYZ.
Please see my explanations in the replies to your comments.

wangyang0918 commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -804,22 +808,31 @@ private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile)
 	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile resourceProfile) {
-		final Collection<ResourceProfile> requestedSlots = resourceActions.allocateResource(resourceProfile);
+		if (workerResourceSpec == null) {
+			// standalone mode, cannot allocate resource
```
I am also curious about why do we make standalone cluster so different? Before this PR, the `StandaloneResourceManager#startNewWorker` is also called and just return an empty collection.

xintongsong replied:
The problem is not about calling `StandaloneResourceManager#startNewWorker`. It is about not having a `WorkerResourceSpec` on standalone setups.

For a standalone cluster, use may have TM resource configurations only on the TM machines. That means trying to create a `WorkerResourceSpec` from the configuration on JM might lead to failures due to missing of memory configurations. I think it does not make sense to force the users to set TM configurations on the JM machine, because they are not really used.

We have run into such problems during the 1.9 release testing, where a standalone JM failed because it tries to calculate a default TM managed memory size.

TisonKun replied:
It should be able to merge into `if (!resourceActions.allocateResource(workerResourceSpec))` branch. But I am not very sure because the next condition is

```
if (!Preconditions.checkNotNull(slotResourceProfile,
			"slotResourceProfile should be null iff workerResourceSpec is null, which means standalone mode.")
			.isMatching(resourceProfile))
```

where we now have three profile(including workerResourceSpec) things become quite complex.

TisonKun replied:
@xintongsong  So where is a document mentions what's the different among `slotResourceProfile`, `resourceProfile` and  `workerResourceSpec`. I can see there are some constraint among them but it is hard to understand from my side.

xintongsong replied:
@TisonKun
Please see my reply to Yang for why calling `allocateResource` on standalone setup. For the same reason, we cannot merge this into the `if (!resourceActions.allocateResource(workerResourceSpec))` branch.

Regarding `slotResourceProfile`, `resourceProfile` and `workerResourceSpec`.
- `slotResourceProfile` describes the resource capacity of a slot.
- `resourceProfile` describes the resource requirement of a slot request.
- `workerResourceSpec` describes the resource requirement of a worker (TM) to request.
I think you're right, these things are indeed hard to understand. How do you think about renaming `resourceProfile` to `requestedSlotResourceProfile`, and `slotResourceProfile` to `defaultSlotResourceProfile`?

TisonKun replied:
@xintongsong thanks for your explanation! That sounds good.

wangyang0918 replied:
Yes, you are right. We should not require the users to set TM resource related configurations on JM side for standalone. This is the right direction. 
What i am thinking is when there are some future changes of the `SlotManagerImpl`, we always need to be careful and keep in mind that `WorkerResourceSpec` should be `null` for standalone cluster.

xintongsong replied:
@wangyang0918
True. That's why we annotate it as `@Nullable` and, as suggested by Tison and updated to the PR, clearly documented that a `null` value means in standalone mode.

TisonKun commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -272,10 +267,9 @@ private void requestKubernetesPod() {
+		final int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		if (requiredTaskManagers > numPendingPodRequests) {
```
I can see the return value of `startWorker` is `@return whether the resource can be allocated`

If so, we should also return boolean in this method where false if we don't request a new pod.

xintongsong replied:
`startNewWorker` is a public interface, and its return value is checked by `SlotManager` for creating pending slots.
For `requestKubernetesPodIfRequired`, I don't see how its return value is needed, at least for now. And since it's a internal private method, I think we can always add that return value if needed later.

TisonKun replied:
Oops I mix up `requestKubernetesPod` with `requestKubernetesPodIfRequired`.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java:
```diff
@@ -0,0 +1,132 @@
+/**
+ * Resource specification of a worker, mainly used by SlotManager requesting from ResourceManager.
+ */
+public class WorkerResourceSpec {
```
There was a discussion about whether reusing `TaskExecutorResourceSpec` in the design google doc. We have decided to not reuse it, and I'll quote what Till commented on the doc here.
> TaskExecutorResourceSpec defines for me the resources of a TaskExecutor. The SlotManager on the other hand is only interested in a certain set of resources which happen to be the exact same set as TaskExecutorResourceSpec offers. However, the latter might evolve adding more resource specifications. Hence, by reusing TaskExecutorResourceSpec we would also couple the SlotManager to the TaskExecutor's resource specification which does not necessarily need to be the same.

wangyang0918 replied:
Thanks for your explanation. Make sense to me.

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnResourceManagerFactory.java:
```diff
@@ -81,4 +91,27 @@ public static YarnResourceManagerFactory getInstance() {
+				cpuCoresDouble,
+				cpuCoresLong);
+		}
+		if (cpuCoresLong > Integer.MAX_VALUE) {
```
I think it is not robust to make the correctness of `YarnResourceManagerFactory` depending on another class `YarnClusterDescriptor`. There is no harm for double checking it, right?

TisonKun replied:
A precondition should work as defensive programming.

xintongsong replied:
@TisonKun Not sure what do you mean. Are you suggesting simplify this code with a `Preconditions.checkstate()`?
I remember the reason we implemented it like this in `YarnResourceManager` at the first place, is to get an `IllegalConfigurationException` instead of a `IllegalStateException`, for better readability to the users.

TisonKun replied:
@xintongsong that makes sense.

wangyang0918 replied:
I am not against with the check. I just feel the `Integer.MAX_VALUE` checking will not take too much effect. Since the `YarnMaxVcores` is far less than `Integer.MAX_VALUE`.

Centainly, you could keep the current check. It is safer when converting `long` to `int`.

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnResourceManagerFactory.java:
```diff
@@ -81,4 +91,27 @@ public static YarnResourceManagerFactory getInstance() {
+		@SuppressWarnings("NumericCastThatLosesPrecision")
+		long cpuCoresLong = Math.max((long) Math.ceil(cpuCoresDouble), 1L);
+		//noinspection FloatingPointEquality
+		if (cpuCoresLong != cpuCoresDouble) {
```
True, we have not exposed `taskmanager.cpu.cores`. But we might in the future, and if we do, it could be easily overlooked to add this check later. Again, I believe there's no harm for over checking it.

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -527,10 +524,9 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		if (requiredTaskManagers > numPendingContainerRequests) {
```
ditto here as in `KubernetesResourceManager `

xintongsong replied:
Same here.

xintongsong commented:
Thanks for the review, @wangyang0918 and @TisonKun.

@wangyang0918, regarding where to create `defaultWorkerResourceSpec`:
- First, it is created in `ActiveResourceManagerFactory`, before creating `SlotManager` and using it to create `ResourceManager`. To that end, I think `ResourceManager` is still not aware of `defaultWorkerResourceSpec`.
- One reason for not creating `defaultWorkerResourceSpec` in `SlotManagerConfiguration`, is that we need to decide whether to create it or not according to the deployment. We should only create it for active RM setups. See my reply on the detailed comment.
- Another reason is that, we need to use different configurations per RM to decide the cpu cores of the `defaultWorkerResourceSpec`.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.Objects;
+
+/**
+ * Resource specification of a worker, mainly used by SlotManager requesting from ResourceManager.
```
suggestion: Document `null` semantic of `WorkerResourceSpec`.

xintongsong replied:
Good point.
I think it might be better to document it in `SlotManagerImpl` rather than `WorkerResourceSpec`. WDYT?

TisonKun replied:
My concern is that we have many `@Nullable` `WorkerResourceSpec` parameters. The semantic of `null` `WorkerResourceSpec` belongs to itself. So we can document the design in the class doc of `WorkerResourceSpec` and linked to this universe doc where a nullable param in use.

xintongsong replied:
I tend to think it differently.
I think the semantic of a `WorkerResourceSpec` being `null` belongs to the context where it is used, not the class itself. E.g., when we create the `SlotManagerImpl`, the `WorkerResourceSpec` means it is on a standalone setup and SM should not request any worker from RM. On the other hand, when we create a `TaskExecutorProcessSpec` from the `WorkerResourceSpce`, it must not be `null`.

TisonKun replied:
Yes that makes sense. Nullable return value and field has their meaning respectively. It seems we have too many Nullable WorkerResourceSpec parameter(just for passing from here to there), maybe they will be later gathered into one single source?

xintongsong replied:
True, we have the `@Nullable WorkerResourceSpec` along the code path for creating the `SlotManagerImpl`. To be specific, in the following classes.
- `ResourceManagerRuntimeServiceConfiguration`
- `SlotManagerConfiguration`
- `SlotManagerImpl`
- `SlotManagerBuilder` (test-only)

I guess the root cause is that we decide `WorkerResourceSpec` differently (whether `null` and the `cpu` config options) across different RMs, so we have to pass it from the `ResourceManagerFactory`. Unfortunately, ATM I don't see a good way to resolve that. Do you have any suggestion on that?

TisonKun replied:
There are too many `XxxConfiguration` which make the call chain much longer. We might 
 refactor to reduce some poor `XxxConfiguration` later. But it doesn't have to be included in this PR.

TisonKun replied:
Or in this pr you can try to directly pass `WorkerResourceSpec` to service's  & manager's `fromConfiguration` instead of configuration's & build's.

xintongsong replied:
I'm not sure why the `XxxConfiguration` is used at the first place. Probably there are some good reasons. From what I see, the `XxxConfiguration` clearly shows which information is needed for creating the `Xxx`.
I think we can discuss the pros and cons of `XxxConfiguration` in another thread, maybe a code style discussion in ML? For this PR, I would also like to exclude it from the scope.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	@VisibleForTesting
+	public WorkerResourceSpec(
```
Share we have a inner Builder to reduce multiple testing purpose constructor override?

xintongsong commented:
@flinkbot run travis
@flinkbot run azure

TisonKun commented:
@flinkbot run azure

xintongsong commented:
Rebased to resolve conflicts.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerFactory.java:
```diff
@@ -81,4 +83,14 @@ private static Configuration createActiveResourceManagerConfiguration(Configurat
 		@Nullable String webInterfaceUrl,
 		ResourceManagerMetricGroup resourceManagerMetricGroup) throws Exception;
+
+	protected WorkerResourceSpec createDefaultWorkerResourceSpec(Configuration configuration) {
```
I'm wondering whether this should really be a method belonging to the `ActiveResourceManagerFactory`. If I look at the body, then it could perfectly be a static method. No need to be method of `ActiveResourceManagerFactory` since it does not access any state.

xintongsong replied:
It calls to the abstract method `getDefaultCpus`, which is currently overridden differently by different `ActiveResourceManagerFactory` implementations.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerFactory.java:
```diff
@@ -81,4 +83,14 @@ private static Configuration createActiveResourceManagerConfiguration(Configurat
+		return WorkerResourceSpec.fromTaskExecutorProcessSpec(taskExecutorProcessSpec);
+	}
+
+	protected abstract CPUResource getDefaultCpus(Configuration configuration);
```
The same applies to this method. I'm not entirely sure whether this class is the right place for this method. The `ActiveResourceManagerFactory` should not need to know how to derive these things ideally.

Not saying that we have to do it right away but I think the cleaner approach would be to pass a `WorkerResourceSpecFactory` to the `ActiveResourceManagerFactory` or even better, pass simply the default `WorkerResourceSpec` to the constructor of the `ActiveResourceManagerFactory`.

xintongsong replied:
If the deriving of cpu cores and default worker spec is too detailed for `ActiveResourceManager`, I would suggest to add an abstract method `getDefaultWorkerSpecFactory` to return different `WorkerResourceSpecFactory` for different `ActiveResourceManager` implementations.

My concern for passing `WorkerResourceSpecFactory` or `WorkerResourceSpec` to `ActiveResourceManager` is that, neither the constructor nor the `createResourceManager` method seems to be a good place to pass them in IMO.
- For the constructor, `ActiveResourceManager` is abstract and its implementations are either singleton (K8s/Yarn) or created by cluster entry point (Mesos). I think it is too detail for the cluster entry point to pass in different `WorkerResourceSpecFactory`/`WorkerResourceSpec` to the resource manager factory, not to mention breaking singleton properties on K8s/Yarn.
- For `createResourceManager`, it is called by `DefaultDispatcherResourceManagerComponentFactory`, which is not aware of the different deployments either.

tillrohrmann replied:
My suggestion would be to pass it to the `ActiveResourceManagerFactory` when creating it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -133,14 +133,17 @@
 			boolean waitResultConsumedBeforeRelease,
-			@Nullable WorkerResourceSpec defaultWorkerResourceSpec) {
+			@Nullable WorkerResourceSpec defaultWorkerResourceSpec,
+			int numSlotsPerWorker) {
```
Maybe it would make sense to pass in the `SlotManagerConfiguration`. If this is possible, then one would not have to add another field to the constructor every time we add a new configuration parameter for the `SlotManager`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -1054,10 +1054,10 @@ protected abstract void internalDeregisterApplication(
 	 */
 	@VisibleForTesting
-	public abstract Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile);
+	public abstract boolean startNewWorker(ResourceProfile resourceProfile);
```
What happens if the RM can only start a multiple of `resourceProfile`? Wouldn't we calculate the number of available slots wrongly in this case?

xintongsong replied:
In this commit, we have moved the logics of calculating number of new pending slots into `SlotManagerImpl`. Therefore, `SlotManagerImpl` only needs to know whether the new worker is started or not, and should be able to correctly calculate the number of available slots.

In the final state, RM accepts `WorkerResourceSpec` instead of slot `ResourceProfile` for starting new workers. It would become a SM decision that how many multiple of `ResourceProfile` the worker should be start with.

tillrohrmann replied:
Ok if I understand it correctly, then this PR relies on the fact that the `RM` and the `SlotManager` both get the same (implicitly or explicitly) `WorkerResourceSpec` assigned, right?

I think it can still happen that the RM won't be able to start exactly a container of the size `WorkerResourceSpec`, even after later changes. Imagine that we are on Yarn were the requested `WorkerResourceSpec` is below the minimum. How should Flink behave in this situation? Should we fail? Should we leave some of the resources unused? Or should we tell the `SlotManager` about it so that it can take the additional resources into account?

xintongsong replied:
True, it is possible that RM won't be able to start a container for given `WorkerResourceSpec`.

I think if the requested resource, to be specific the `TaskExecutorProcessSpec` derived from `WorkerResourceSpec`, is smaller than the Yarn min-allocation, we can simply throw a warning for that. Same for when requested resource is not an integer multiple of min-allocation. The critical part is if the requested resource is larger than the max-allocation, and we should check return false in such cases.

I think these checks should be performed by the RM implementations. I'll add the checks for Yarn in #11353. For K8s, I think there are no such min/max-allocation limits, but I'll double-check on that.

tillrohrmann replied:
Hmm, but with this API we will waste resources because the caller thinks it has started a worker of the specified size, right? An idea how to improve the situation is to let the `startNewWorker` method return the size of the container it has started. This would allow the caller to react to it.

xintongsong replied:
Looking more into this, I'm not sure whether returning the container size is the right approach.

The problem is, the container resource contains two parts that are decided by different components.
- Those contained in `WorkerResourceSpec` are decided by `SlotManager`: cpu, task heap, task off-heap, network, managed.
- Those contained in `TaskExecutorProcessSpec` but not in `WorkerResourceSpec` are decided by `ResourceManager`: framework heap, framework off-heap, jvm metaspace, jvm overhead.

If we return the size of the entire started container to SM, it does not help SM to decide whether there are resources wasted, because SM is not aware of how many resources in addition to `WorkerResourceSpec` that the framework and jvm need.

An alternative is to exclude the framework and jvm resources and the remaining to SM. The problem is that, the excluded part is not constant and might be dependent on `WorkerResourceSpec`. To be specific, jvm overhead might be configured as a fraction of total process memory size. That means increasing the total size of `WorkerResourceSpec` may also lead to the increasing of jvm overhead.

I think what should be returned by this API really depends on how we want to solve the resource wasting issue. TBH, I'm not sure when would we have a `SlotManager` implementation that deals with the Yarn resource wasting problem. I would suggest to create a follow-up ticket and leave it to when we do have such a SM implementation. WDYT?

tillrohrmann replied:
Sounds good to me. Thanks for the background information.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -153,6 +155,8 @@ public SlotManagerImpl(
 		this.waitResultConsumedBeforeRelease = waitResultConsumedBeforeRelease;
 		this.defaultWorkerResourceSpec = defaultWorkerResourceSpec;
 		this.numSlotsPerWorker = numSlotsPerWorker;
+		this.defaultSlotResourceProfile = defaultWorkerResourceSpec != null ?
+			generateDefaultSlotResourceProfile(defaultWorkerResourceSpec, numSlotsPerWorker) : null;
```
Instead of setting `defaultSlotResourceProfile` to `null`, can't we set it to `ResourceProfile.UNKNOWN` or something like this?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -807,23 +811,32 @@ private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile)
-	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile resourceProfile) {
-		final Collection<ResourceProfile> requestedSlots = resourceActions.allocateResource(resourceProfile);
+	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile requestedSlotResourceProfile) {
+		if (defaultWorkerResourceSpec == null) {
+			// standalone mode, cannot allocate resource
+			return Optional.empty();
+		}
```
I think here we are leaking details from the enclosing component into the `SlotManager`. I think the `SlotManager` should not decide for the `ResourceManager` whether it can allocate new resources or not.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -807,23 +811,32 @@ private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile)
-			final Iterator<ResourceProfile> slotIterator = requestedSlots.iterator();
-			final PendingTaskManagerSlot pendingTaskManagerSlot = new PendingTaskManagerSlot(slotIterator.next());
-			pendingSlots.put(pendingTaskManagerSlot.getTaskManagerSlotId(), pendingTaskManagerSlot);
+		}
```
Shouldn't we instead of returning an `Optional.empty()` fail with an invalid state exception because the current `SlotManagerImpl` does not support custom resource profiles yet?

xintongsong replied:
I'm not sure whether this should be an invalid state exception. 

It is true that all the slot requests should have unknown resource profile ATM. However, I think `SlotManagerImpl` does not necessarily need to know whether there're custom profiles or not. As long as the requested resource profile can be matched by the default slot profile, it should be able to allocate the slot. And if the requested slot profile is larger than the default slot, an `UnfulfillableSlotRequestException` should be thrown.

tillrohrmann replied:
Failing in case that the requested slot profile is too large sound good to me.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceActions.java:
```diff
@@ -42,9 +40,9 @@
+	 * @return whether the resource can be allocated
 	 */
-	Collection<ResourceProfile> allocateResource(ResourceProfile resourceProfile);
+	boolean allocateResource(ResourceProfile resourceProfile);
```
Wouldn't it be required for future developments that we can tell `allocateResource` how big resource-wise the worker should be?

xintongsong replied:
I'm not sure whether I have correctly understood your question.
Telling `allocateResource` how big the worker is in the scope of making slot allocation pluggable. In this PR, we have the SM to tell `allocateResource` the `WorkerResourceSpec`, in a subsequence commit rather than this one. This PR does not make the RM implementations to respect the `WorkerResourceSpec` they are told. We make those changes in subsequence PRs.

tillrohrmann replied:
All right, understood.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerFailUnfulfillableTest.java:
```diff
@@ -52,6 +53,9 @@
  */
 public class SlotManagerFailUnfulfillableTest extends TestLogger {
 
+	private static final WorkerResourceSpec WORKER_RESOURCE_SPEC = new WorkerResourceSpec(
+		100.0, 10000, 10000, 10000, 10000);
```
Looking at this line, I would recommend introducing a builder for the `WorkerResourceSpec` and to make the default constructor private.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -267,10 +267,9 @@ private void requestKubernetesPod() {
+		final int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		if (requiredTaskManagers > numPendingPodRequests) {
```
Unrelated. Wouldn't a while loop be better here?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -1206,8 +1206,12 @@ public Void retrievePayload(ResourceID resourceID) {
-	protected int getNumberRequiredTaskManagerSlots() {
-		return slotManager.getNumberPendingTaskManagerSlots();
+	protected int getNumberRequiredTaskManagers() {
+		return getPendingWorkerNums().values().stream().reduce(0, Integer::sum);
```
For the time being I would add a check state which ensures that the `WorkerResourceSpecs` can all be fulfilled by the container size we are using to start new containers/pods.

xintongsong replied:
IIUC, this is because the RM implementations are ignoring the `WorkerResourceSpec` passed to `startNewWorker`? Then would it be better to add this check in `startNewWorker`?

tillrohrmann replied:
If this is the entry point, then we can do the check there. The important bit is to maintain the invariant that we do not start containers for `WorkerResourceSpecs` which are larger than the container.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -51,7 +53,7 @@
 	int getNumberFreeSlotsOf(InstanceID instanceId);
 
-	int getNumberPendingTaskManagerSlots();
+	Map<WorkerResourceSpec, Integer> getPendingWorkerNums();
```
`JavaDoc` missing.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -207,6 +208,18 @@ public int getNumberFreeSlotsOf(InstanceID instanceId) {
+		final int pendingWorkerNum = (int) Math.ceil((double) pendingSlots.size() / numSlotsPerWorker);
+		if (pendingWorkerNum > 0) {
+			return Collections.singletonMap(
+				Preconditions.checkNotNull(defaultWorkerResourceSpec,
+					"There should never be pending slots/workers in standalone mode."),
```
I think this should not be necessary.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -524,10 +524,9 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		if (requiredTaskManagers > numPendingContainerRequests) {
```
Same here with the `while` loop. But also unrelated to this PR as the other comment.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -156,8 +156,8 @@ protected void internalDeregisterApplication(ApplicationStatus finalStatus, @Nul
-		LOG.info("Starting new worker with resource profile, {}", resourceProfile);
+	public boolean startNewWorker(WorkerResourceSpec workerResourceSpec) {
+		LOG.info("Starting new worker with worker resource spec, {}", workerResourceSpec);
 		requestKubernetesPod();
```
I guess that we need a check that the configured pod fulfills `workerResourceSpec`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -1051,13 +1050,13 @@ protected abstract void internalDeregisterApplication(
+	 * Allocates a resource using the worker resource specification.
 	 *
-	 * @param resourceProfile The resource description
+	 * @param workerResourceSpec that describes the to be allocated worker.
```
```suggestion
	 * @param workerResourceSpec that describes the to be allocated worker
```

param, return or throws tags are usually not terminated by a period.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceActions.java:
```diff
@@ -37,12 +37,12 @@
+	 * Requests to allocate a resource with the given {@link WorkerResourceSpec}.
 	 *
-	 * @param resourceProfile for the to be allocated resource
+	 * @param workerResourceSpec for the to be allocated worker
```
Usually `@param` requires a noun phrase: `@param workerResourceSpec workerResourceSpec specifies the size of the to be allocated resource`.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -288,7 +288,7 @@ protected void internalDeregisterApplication(
 
 	@Override
-	public boolean startNewWorker(ResourceProfile resourceProfile) {
+	public boolean startNewWorker(WorkerResourceSpec workerResourceSpec) {
```
I think we should ensure that `workerResourceSpec` can be fulfilled by the actual container size the `YarnResourceManager` starts.

tillrohrmann commented:
I think the last part of my comment was answered by #11353.

xintongsong commented:
Thanks for the review, @tillrohrmann.

Regarding the nullable `defaultWorkerResourceSpec` and `defaultSlotResourceProfile`, I'm trying to understand and summarize your concerns as follows. Please correct me if I'm wrong.
- It is not explicit what the null values means and should be handled.
- It makes `SlotManagerImpl` kind of inferring which deployment (standalone or active) it works on, and behave differently.
These concerns make good sense to me.

Actually, I have also considered having a special `WorkerResourceSpec.UNKNOWN`. However, I think this is different from `ResourceProfile.UNKNOWN` or `ResourceSpec.UNKNOWN`. For `ResourceProfile`/`ResourceSpec`, `UNKNOWN` means we need some resource but we don't known how many exactly. For `defaultWorkerResourceSpec`, the semantic we actually need for standalone clusters is "the default worker does not exist", or in other words "by default we do not request workers".

My concern is that, `WorkerResourceSpec` is also used as the argument type of `ResourceActions#allocateResource`. With the current design, SM should always request workers from RM with specific worker resource spec. If we add WorkerResourceSpec.UNKNOWN, we have to carefully make sure it is never passed to `allocateResource`.

Combining your concerns and mine, I would propose the following approach.
- Introduce an `DefaultWorkerSlotResourceProvider`, with methods `Optional<WorkerResourceSpec> getDefaultWorkerResourceSpec` and `Optional<ResourceProfile> getDefaultSlotResourceProfile`.
- For standalone clusters, we can pass a special `DefaultWorkerSlotResourceProvider.NO_DEFAULTS` to `SlotManagerImpl`, which returns `Optional.empty()` for both methods. The absent optional value clearly reflects the semantics of "not existing".
- `SlotManagerImpl` no longer needs to assume which RM implementation it works with. It simply not allocate resource from RM if `DefaultWorkerSlotResourceProvider#getDefaultWorkerResourceSpec` returns empty.

WDYT?

tillrohrmann commented:
I see your point @xintongsong and agree that the caller (`SlotManager`) should not call the `RM` with some invalid value. I guess what we would really need is an `ActiveSlotManager` and a `SlotManager` which cannot and won't try to allocate new resources.

Why is it that the standalone `SlotManager` cannot generate a `WorkerResourceSpec`? Shouldn't we be able to calculate it? Given that the actual value should not matter, we could also default to some random/arbitrary value.

If this is not possible, then I'm ok with adding a `DefaultWorkerSlotResourceProvider` and having a `DefaultWorkerSlotResourceProvider.NO_RESOURCES` or so.

xintongsong commented:
@tillrohrmann 

My concern for standalone clusters generating `WorkerResourceSpec` was that, it requires proper memory configurations and may fail if certain options are absent or conflicts exist. For a standalone cluster, user may only have the memory configurations on the TM machines but not on the JM machine. It is not a good experience to fail the JM process due to the calculation of `WorkerResourceSpec` which is not really needed in such scenarios.

I think you're right, a random/arbitrary `WorkerResourceSpec` should help reduce the complexity. It should not cause any problem since the `StandaloneResourceManager` will ignore the requested `WorkerResourceSpec` anyway.

Thanks again for the review and suggestions. I'll try to address them and update the PR asap.

tillrohrmann commented:
I think you are right that the standalone `Dispatcher` should not fail if it cannot create the `WorkerResourceSpec` because there are some conflicting entries.

Ok, then let's go with some arbitrary/default worker spec for the standalone case.

xintongsong commented:
@tillrohrmann 
I've addressed your comments, except for one about returning size of the started container to slot manager. Please see the corresponding comment thread for details.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -267,10 +267,9 @@ private void requestKubernetesPod() {
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		while (requiredTaskManagers-- > numPendingPodRequests) {
```
```suggestion
		while (requiredTaskManagers > numPendingPodRequests) {
```
I think `requestKubernetesPod` will increase `numPendingPodRequests`.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -267,10 +267,9 @@ private void requestKubernetesPod() {
 	private void requestKubernetesPodIfRequired() {
-		final int requiredTaskManagerSlots = getNumberRequiredTaskManagerSlots();
-		final int pendingTaskManagerSlots = numPendingPodRequests * numSlotsPerTaskManager;
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
```
```suggestion
		final int requiredTaskManagers = getNumberRequiredTaskManagers();
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -1206,8 +1206,12 @@ public Void retrievePayload(ResourceID resourceID) {
+	}
+
+	protected Map<WorkerResourceSpec, Integer> getPendingWorkerNums() {
+		return slotManager.getPendingWorkerNums();
```
Should we rename this method into `getRequiredResources` or so?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -51,7 +53,12 @@
+	 * @return a map whose key set is all the unique resource specs of the pending workers,
+	 * and the corresponding value is number of pending workers of that resource spec.
+	 */
+	Map<WorkerResourceSpec, Integer> getPendingWorkerNums();
```
```suggestion
	Map<WorkerResourceSpec, Integer> getRequiredResources();
```
maybe?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -199,6 +200,14 @@ public int getNumberFreeSlotsOf(InstanceID instanceId) {
 
 	@Override
+	public Map<WorkerResourceSpec, Integer> getPendingWorkerNums() {
+		final int pendingWorkerNum = (int) Math.ceil((double) pendingSlots.size() / numSlotsPerWorker);
```
If we want to get rid of back and forth casting and double calculations:
```suggestion
		final int pendingWorkerNum = (pendingSlots.size() - 1) / numSlotsPerWorker + 1;
```

tillrohrmann replied:
On the other hand, it complicates the expression a bit. Maybe introducing a name for this function `multipleOfAndLargerOrEqualThan(numSlotsPerWorker, pendingSlots.size())`.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -524,10 +524,9 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
 	private void requestYarnContainerIfRequired() {
-		int requiredTaskManagerSlots = getNumberRequiredTaskManagerSlots();
-		int pendingTaskManagerSlots = numPendingContainerRequests * numSlotsPerTaskManager;
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
```
```suggestion
		final int requiredTaskManagers = getNumberRequiredTaskManagers();
```

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -524,10 +524,9 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
+		int requiredTaskManagers = getNumberRequiredTaskManagers();
 
-		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+		while (requiredTaskManagers-- > numPendingContainerRequests) {
```
```suggestion
		while (requiredTaskManagers > numPendingContainerRequests) {
```

xintongsong commented:
Thanks @tillrohrmann. Comments addressed.

xintongsong commented:
Rebased onto the latest master, and added fix for PerJobMiniClusterFactoryTest.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterFactory.java:
```diff
@@ -100,8 +100,10 @@ private MiniClusterConfiguration getMiniClusterConfig(int maximumParallelism) {
-		int numSlotsPerTaskManager = configuration.getInteger(TaskManagerOptions.NUM_TASK_SLOTS, maximumParallelism);
+		int numSlotsPerTaskManager = configuration.getOptional(TaskManagerOptions.NUM_TASK_SLOTS)
+			.orElseGet(() -> maximumParallelism > 0 ?
+				(int) Math.ceil((double) maximumParallelism / numTaskManagers) :
```
Can't we use `MathUtils.divideRoundUp` here?

tillrohrmann commented:
Thanks @xintongsong for fixing the test. I had one last comment concerning this hotfix.

xintongsong commented:
@tillrohrmann 
True, I cherry-picked this hotfix commit from another branch, where `MathUtils.divideRoundU` was not introduced.
I've updated the PR.

tillrohrmann commented:
All right, thanks for updating this PR. I'll take it from here and merge it once Travis gives green light.

xintongsong commented:
Thanks a lot, Till~
